forge-ai/src/main/java/forge/ai/PlayerControllerAi.java

```diff
@@ -105,8 +102,98 @@ public boolean isAI() {
+        int sbChanceOnWin = getAi().getIntProperty(AiProps.SIDEBOARDING_CHANCE_ON_WIN);
+        int sbChancePerCard = getAi().getIntProperty(AiProps.SIDEBOARDING_CHANCE_PER_CARD);
+
+        if (!sbLimitedFormats && gameType.isCardPoolLimited()) {
+            return null;
+        }
+
+        if (MyRandom.getRandom().nextInt(100) < sbChanceOnWin) {
+            GameOutcome lastOutcome = brains.getGame().getMatch().getOutcomeById(brains.getGame().getId() - 1);
+            if (lastOutcome.getWinningPlayer().getPlayer().equals(player.getLobbyPlayer())) {
+                return null;
+            }
+        }
+
+        List<PaperCard> main = deck.get(DeckSection.Main).toFlatList();
+        List<PaperCard> sideboard = deck.get(DeckSection.Sideboard).toFlatList();
+
+        // Devise a sideboarding plan
+        // TODO: Allow a pre-planned (e.g. fully transformative) sideboard via deck metadata of some kind?
+        List<PaperCard> processed = Lists.newArrayList();
+        Map<PaperCard, PaperCard> sideboardPlan = Maps.newHashMap();
+        for (PaperCard cSide : sideboard) {
+            if (processed.contains(cSide)) {
+                continue;
+            } else if (cSide.getRules().getAiHints().getRemAIDecks()) {
+                continue; // don't sideboard in anything that we don't know how to play
+            } else if (cSide.getRules().getType().isLand()) {
+                continue; // don't know how to sideboard lands efficiently yet
+            }
+
+            for (PaperCard cMain : main) {
+                if (processed.contains(cMain)) {
+                    continue;
+                } else if (cMain.equals(cSide)) {
+                    continue;
+                } else if (cMain.getRules().getType().isLand()) {
+                    continue; // don't know how to sideboard lands efficiently yet
+                }
+
+                if (sbSharedTypesOnly) {
+                    if (!cMain.getRules().getType().sharesCardTypeWith(cSide.getRules().getType())) {
+                        continue; // Only equivalent types allowed
+                    }
+                } else {
+                    if ((cMain.getRules().getType().isCreature() && !cSide.getRules().getType().isCreature())
+                            || (cSide.getRules().getType().isCreature()) && !cMain.getRules().getType().isCreature()) {
+                        if (!(sbPlaneswalkerException && (cMain.getRules().getType().isPlaneswalker() || cSide.getRules().getType().isPlaneswalker()))) {
+                            continue; // Creature exception: only trade a creature for another creature unless planeswalkers are allowed as a replacement
+                        }
+                    }
+                }
+
+                if (!Card.fromPaperCard(cMain, player).getManaAbilities().isEmpty()) {
+                    continue; // Mana Ability exception: Don't sideboard out cards that produce mana, can screw up the mana base
```

Comment:
also add it to processed then so you don't have to perform this expensive check more than once?